[Navigation] Null Pointer in NavigationImpl if navigation root missing jcr:content

## Bug Report

**Current Behavior**
If a navigation component has its navigation root property set to a page, and that page does not have a jcr:content child, then a NPE is thrown. This situation occurs sometimes when an author does not publish a parent page before publishing a child page. Although it is the result of user error, i do not believe that it should result in a 500 error.

```
Caused by: java.lang.NullPointerException: null
	at com.day.cq.wcm.core.impl.LanguageManagerImpl.getLanguageRootPath(LanguageManagerImpl.java:499) [com.day.cq.wcm.cq-wcm-core:5.11.276]
	at com.day.cq.wcm.core.impl.LanguageManagerImpl.getLanguageRootResource(LanguageManagerImpl.java:554) [com.day.cq.wcm.cq-wcm-core:5.11.276]
	at com.day.cq.wcm.core.impl.LanguageManagerImpl.getLanguageRoot(LanguageManagerImpl.java:537) [com.day.cq.wcm.cq-wcm-core:5.11.276]
	at com.day.cq.wcm.core.impl.LanguageManagerImpl.getLanguageRoot(LanguageManagerImpl.java:529) [com.day.cq.wcm.cq-wcm-core:5.11.276]
	at com.adobe.cq.wcm.core.components.internal.models.v1.NavigationImpl.getItems(NavigationImpl.java:106) [com.adobe.cq.core.wcm.components.core:2.5.0]
```

**Expected behavior/code**
Not a 500 error.

**Environment**
- AEM 6.4
- Core Components version 2.5.0
- JRE version 1.8.0

**Possible Solution**
Just wrapping the call in a null check would prevent the NPE.
However, the #getLanguageRootPath() method doesn't actually need the page's content resource, it should work fine with the resource for the actual page it's self (i.e. parent of the content resource). The resource located at the page path of the current page is very likely to exist.

--- a/bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/models/v1/NavigationImpl.java
+++ b/bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/models/v1/NavigationImpl.java
@@ -19,6 +19,7 @@ import java.util.ArrayList;
 import java.util.Collections;
 import java.util.Iterator;
 import java.util.List;
+import java.util.Optional;
 import javax.annotation.PostConstruct;
 import javax.jcr.RangeIterator;
 
@@ -107,7 +108,7 @@ public class NavigationImpl implements Navigation {
             Page rootPage = pageManager.getPage(navigationRootPage);
             if (rootPage != null) {
                 NavigationRoot navigationRoot = new NavigationRoot(rootPage, structureDepth);
-                Page navigationRootLanguageRoot = languageManager.getLanguageRoot(navigationRoot.page.getContentResource());
+                Page navigationRootLanguageRoot = navigationRoot.getPageResource().map(languageManager::getLanguageRoot).orElse(null);
                 Page currentPageLanguageRoot = languageManager.getLanguageRoot(currentPage.getContentResource());
                 RangeIterator liveCopiesIterator = null;
                 try {
@@ -226,7 +227,7 @@ public class NavigationImpl implements Navigation {
     }
 
     private class NavigationRoot {
-        Page page;
+        final Page page;
         int startLevel;
         int structureDepth = -1;
 
@@ -236,6 +237,22 @@ public class NavigationImpl implements Navigation {
             if (configuredStructureDepth > -1) {
                 structureDepth = configuredStructureDepth + startLevel;
             }
+        }
+
+        /**
+         * Gets the resource representation of the navigation root page.
+         *
+         * @return The resource for the navigation root, empty if the resource could not be resolved.
+         */
+        @NotNull
+        final Optional<Resource> getPageResource() {
+            return Optional.ofNullable(
+                Optional.of(this.page)
+                    // get the parent of the content resource
+                    .map(Page::getContentResource)
+                    .map(Resource::getParent)
+                    // if content resource is missing, resolve resource at page path
+                    .orElseGet(() -> resourceResolver.getResource(this.page.getPath())));
         }
     }
 

--- a/bundles/core/src/test/java/com/adobe/cq/wcm/core/components/internal/models/v1/NavigationImplTest.java
+++ b/bundles/core/src/test/java/com/adobe/cq/wcm/core/components/internal/models/v1/NavigationImplTest.java
@@ -78,6 +78,9 @@ public class NavigationImplTest {
     private static final String NAV_COMPONENT_9 = TEST_ROOT + "/jcr:content/root/navigation-component-9";
     // points to the nav component used for invalidRedirectTest()
     private static final String NAV_COMPONENT_10 = TEST_ROOT + "/jcr:content/root/navigation-component-10";
+    // points to the nav component used for when the nav root has no jcr:content child
+    private static final String NAV_COMPONENT_11 = TEST_ROOT + "/jcr:content/root/navigation-component-11";
+
 
     private static final ContentPolicyManager contentPolicyManager = mock(ContentPolicyManager.class);
 
@@ -162,6 +165,20 @@ public class NavigationImplTest {
         Navigation navigation = getNavigationUnderTest(NAV_COMPONENT_2);
         assertEquals("Didn't expect any navigation items.", 0, navigation.getItems().size());
         Utils.testJSONExport(navigation, Utils.getTestExporterJSONPath(TEST_BASE, "navigation4"));
+    }
+
+    /**
+     * Demonstrates the ability to construct a {@link NavigationImpl} where the navigation root page does not have a
+     * jcr:content node, but does have legitimate sub-pages.
+     */
+    @Test
+    public void testNavigationRootMissingJCRContent() {
+        Navigation navigation = getNavigationUnderTest(NAV_COMPONENT_11);
+        Object[][] expectedPages = {
+            {"/content/navigation-missing-jcr-content/navigation-1", 0, false, "/content/navigation-missing-jcr-content/navigation-1.html"},
+            {"/content/navigation-missing-jcr-content/navigation-2", 0, false, "/content/navigation-missing-jcr-content/navigation-2.html"}
+        };
+        verifyNavigationItems(expectedPages, navigation.getItems());
     }
 
     @Test

--- a/bundles/core/src/test/resources/navigation/test-content.json
+++ b/bundles/core/src/test/resources/navigation/test-content.json
@@ -24,6 +24,12 @@
                     "sling:resourceType": "core/wcm/components/navigation/v1/navigation",
                     "navigationRoot"    : "/content/navigation-invalid-redirect",
                     "skipNavigationRoot": false
+                },
+                "navigation-component-11": {
+                    "jcr:primaryType"   : "nt:unstructured",
+                    "sling:resourceType": "core/wcm/components/navigation/v1/navigation",
+                    "navigationRoot"    : "/content/navigation-missing-jcr-content",
+                    "skipNavigationRoot": true
                 }
             }
         },
@@ -470,6 +476,25 @@
                         "skipNavigationRoot": false
                     }
                 }
+            }
+        }
+    },
+    "navigation-missing-jcr-content"         : {
+        "jcr:primaryType": "cq:Page",
+        "navigation-1"   : {
+            "jcr:primaryType": "cq:Page",
+            "jcr:content"    : {
+                "jcr:primaryType" : "cq:PageContent",
+                "jcr:title"       : "Navigation 1",
+                "jcr:created"     : "Thu Jun 29 2017 12:13:52 GMT+0200"
+            }
+        },
+        "navigation-2"   : {
+            "jcr:primaryType": "cq:Page",
+            "jcr:content"    : {
+                "jcr:primaryType": "cq:PageContent",
+                "jcr:title"      : "Navigation 2",
+                "jcr:created"    : "Thu Jun 29 2017 12:13:52 GMT+0200"
             }
         }
     }

